### ISSUE TITLE
Pylance reports "Method not defined" when lru_cache() decorator added

### ISSUE THREAD

**Issue by Shapedsundew9** [microsoft/pylance-release#5314]
## Issue
Adding lru_cache() function decorator confuses pylance method resolution.
![image](https://github.com/microsoft/pylance-release/assets/16618209/62f511d8-447a-4c0c-9fc8-20a5003f8f57)

## Environment data

```
Distributor ID:	Ubuntu
Description:	Ubuntu 23.10
Release:	23.10
Codename:	mantic
```

```
2024-01-01 16:07:47.310 [info] (Client) Pylance async client (2023.12.1) started with python extension (2023.22.1)
2024-01-01 16:07:48.151 [info] [Info  - 4:07:48 PM] (420204) Pylance language server 2023.12.1 (pyright a78e2b10) starting
2024-01-01 16:07:48.151 [info] [Info  - 4:07:48 PM] (420204) Server root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.12.1/dist
2024-01-01 16:07:48.159 [info] [Info  - 4:07:48 PM] (420204) Starting service instance "scratch"
2024-01-01 16:07:48.221 [info] [Info  - 4:07:48 PM] (420204) Setting pythonPath for service "scratch": "/home/shapedsundew9/Projects/.venv/bin/python"
2024-01-01 16:07:48.221 [info] [Info  - 4:07:48 PM] (420204) Setting environmentName for service "scratch": "3.11.6 (.venv venv)"
2024-01-01 16:07:48.283 [info] [Info  - 4:07:48 PM] (420204) Assuming Python version 3.11
2024-01-01 16:07:48.398 [info] [Info  - 4:07:48 PM] (420204) Found 1 source file
```

## Code Snippet

```python
"""Example"""
from typing import Self
from functools import lru_cache


class A:
    """A class"""
    def __getitem__(self, _) -> Self:
        return self


class B(A):
    """B class"""
    @lru_cache()
    def __getitem__(self, _) -> Self:
        return self


a: A = A()[0]
b: B = B()[0]
```

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported. Please try creating a new project and copying & pasting the above code to confirm that you're able to repro the issue.

My only theory is that the bundled `functools.pyi` type stub was corrupted or modified on your system. Do you perhaps recall making edits to that file? You might need to uninstall and reinstall pylance if that file is corrupt.

**Comment by debonte** [comment]
I believe this is a dupe of https://github.com/microsoft/pyright/issues/6500 which was fixed in Pyright 1.1.337. That fix was included in Pylance 2023.12.101 (latest prerelease), so you can eliminate the issue by switching to Pylance prerelease.
